***Homewrok 2 (model 2, lesson 2)***

In [64]:
import pandas as pd
import numpy as np
url = " https://raw.githubusercontent.com/leonism/sample-superstore/master/data/superstore.csv"
try:
    df = pd.read_csv(url, encoding='latin-1')
    print(f"Loaded from URL: {df.shape[0]} rows, {df.shape[1]} columns")
except Exception as e:
    print(f"URL failed ({e}). Use Option A: upload the CSV manually.")
    # Fallback: upload manually
    from google.colab import files
    uploaded = files.upload()
    filename = list(uploaded.keys())[0]
    df = pd.read_csv(filename, encoding='latin-1')
    print(f"Loaded from upload: {df.shape[0]} rows, {df.shape[1]} columns")

Loaded from URL: 10800 rows, 21 columns


In [65]:
# shows rows and columns
print("Shape:", df.shape)

Shape: (10800, 21)


In [66]:
df.shape

(10800, 21)

In [67]:
# non-null counts - not filled columns
# df.info() - shows non-null count and Dtype
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10800 entries, 0 to 10799
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Row ID         10800 non-null  object 
 1   Order ID       10800 non-null  object 
 2   Order Date     9994 non-null   object 
 3   Ship Date      9994 non-null   object 
 4   Ship Mode      9994 non-null   object 
 5   Customer ID    9994 non-null   object 
 6   Customer Name  9994 non-null   object 
 7   Segment        9994 non-null   object 
 8   Country        9994 non-null   object 
 9   City           9994 non-null   object 
 10  State          9994 non-null   object 
 11  Postal Code    9983 non-null   float64
 12  Region         9994 non-null   object 
 13  Product ID     9994 non-null   object 
 14  Category       9994 non-null   object 
 15  Sub-Category   9994 non-null   object 
 16  Product Name   9994 non-null   object 
 17  Sales          9994 non-null   float64
 18  Quanti

In [68]:
#df.describe() -  for summary statistics of numerical columns (shows count, mean, std, min, max ... of data)
df.describe()

,Postal Code,Sales,Quantity,Discount,Profit
count,9983.000000,9994.000000,9994.000000,9994.000000,9994.000000
mean,55245.233297,229.858001,3.789574,0.156203,28.656896
std,32038.715955,623.245101,2.225110,0.206452,234.260108
min,1040.000000,0.444000,1.000000,0.000000,-6599.978000
25%,23223.000000,17.280000,2.000000,0.000000,1.728750
50%,57103.000000,54.490000,3.000000,0.200000,8.666500
75%,90008.000000,209.940000,5.000000,0.200000,29.364000
max,99301.000000,22638.480000,14.000000,0.800000,8399.976000


In [69]:
#Check for missing values
print(df.isnull().sum())

Row ID             0
Order ID           0
Order Date       806
Ship Date        806
Ship Mode        806
Customer ID      806
Customer Name    806
Segment          806
Country          806
City             806
State            806
Postal Code      817
Region           806
Product ID       806
Category         806
Sub-Category     806
Product Name     806
Sales            806
Quantity         806
Discount         806
Profit           806
dtype: int64


The dataset contains 10,800 rows and 21 columns. There are two data types: float64 and object. All columns have missing values except for Order ID and Name ID. The Postal Code column has 817 missing values, while the remaining columns each have 806 missing values.

The Postal Code and Discount columns show a left-skewed distribution (i.e., the data contains many large values and a few small ones).

The Sales, Quantity, and Profit columns exhibit a right-skewed distribution (i.e., the data contains many small values and a few large ones, including outliers).

In [70]:
# For numerical columns, fill with median (more robust than mean if outliers exist)
df['Sales'] = df['Sales'].fillna(df['Sales'].median())
df['Postal Code'] = df['Postal Code'].fillna(df['Postal Code'].median())
df['Quantity'] = df['Sales'].fillna(df['Quantity'].median())
df['Discount'] = df['Discount'].fillna(df['Discount'].median())
df['Profit'] = df['Profit'].fillna(df['Profit'].median())
#  For categorical columns, fill with mode (most frequent value)
df['Ship Mode'] = df['Ship Mode'].fillna(df['Ship Mode'].mode()[0])
df['Segment'] = df['Segment'].fillna(df['Segment'].mode()[0])
df['Region'] = df['Region'].fillna(df['Region'].mode()[0])
df['Category'] = df['Category'].fillna(df['Category'].mode()[0])
df['Sub-Category'] = df['Sub-Category'].fillna(df['Sub-Category'].mode()[0])
df['State'] = df['State'].fillna(df['State'].mode()[0])
df['City'] = df['City'].fillna(df['City'].mode()[0])
df['Country'] = df['Country'].fillna(df['Country'].mode()[0])
df['Order Date'] = df['Order Date'].fillna(df['Order Date'].mode()[0])
df['Ship Date'] = df['Ship Date'].fillna(df['Ship Date'].mode()[0])
df['Customer ID'] = df['Customer ID'].fillna(df['Customer ID'].mode()[0])
df['Customer Name'] = df['Customer Name'].fillna(df['Customer Name'].mode()[0])
df['Product ID'] = df['Product ID'].fillna(df['Product ID'].mode()[0])
df['Product Name'] = df['Product ID'].fillna(df['Product Name'].mode()[0])

In [71]:
# Verify no missing values remain:
print(df.isnull().sum())

Row ID           0
Order ID         0
Order Date       0
Ship Date        0
Ship Mode        0
Customer ID      0
Customer Name    0
Segment          0
Country          0
City             0
State            0
Postal Code      0
Region           0
Product ID       0
Category         0
Sub-Category     0
Product Name     0
Sales            0
Quantity         0
Discount         0
Profit           0
dtype: int64


All columns with the object data type were filled using the mode, and columns with the float data type were also filled using the mode. Overall, 806 missing values were filled, except for Row ID and Order ID, which had no missing values. “Median is robust to outliers, while mean is sensitive.” “Mode is suitable for categorical variables because it represents the most frequent category.”

In [72]:
# count duplicates
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 504


In [73]:
#  Remove them
df = df.drop_duplicates()
print("Shape after removing duplicates:", df.shape)

Shape after removing duplicates: (10296, 21)


There are 504 duplicate rows. After removing them, the dataset contains 10,296 rows.

In [74]:
# Convert the date columns
# Order Date column → was text (string) → datetime
# Ship Date column →  was text (string) → datetime
df['Order Date'] = pd.to_datetime(df['Order Date'])
df['Ship Date'] = pd.to_datetime(df['Ship Date'])

In [75]:
# Verify the conversion
print(df.dtypes)

Row ID                   object
Order ID                 object
Order Date       datetime64[ns]
Ship Date        datetime64[ns]
Ship Mode                object
Customer ID              object
Customer Name            object
Segment                  object
Country                  object
City                     object
State                    object
Postal Code             float64
Region                   object
Product ID               object
Category                 object
Sub-Category             object
Product Name             object
Sales                   float64
Quantity                float64
Discount                float64
Profit                  float64
dtype: object


In [76]:
#  Spot-check a few values to make sure dates look correct
df[['Order Date', 'Ship Date']].head()

,Order Date,Ship Date
0,2017-11-08,2017-11-11
1,2017-11-08,2017-11-11
2,2017-06-12,2017-06-16
3,2016-10-11,2016-10-18
4,2016-10-11,2016-10-18


In [80]:
# Calculate total spending per customer:
total_spending = df.groupby('Customer ID')['Sales'].sum().reset_index()
total_spending.columns = ['Customer ID', 'total_spending']

In [78]:
#  Calculate order frequency (number of orders per customer:
order_freq = df.groupby('Customer ID')['Order ID'].nunique().reset_index()
order_freq.columns = ['Customer ID', 'order_frequency']

In [82]:
# Calculate average order value per customer:
avg_order = df.groupby('Customer ID')['Sales'].mean().reset_index()
avg_order.columns = ['Customer ID', 'avg_order_value']

# Merge all into a customer summary table:
customer_summary = total_spending.merge(order_freq, on='Customer ID').merge(avg_order, on='Customer ID')
customer_summary.head(10)

,Customer ID,total_spending,order_frequency,avg_order_value
0,AA-10315,5563.560,5,505.778182
1,AA-10375,1056.390,9,70.426000
2,AA-10480,1790.512,4,149.209333
3,AA-10645,5086.935,6,282.607500
4,AB-10015,886.156,3,147.692667
5,AB-10060,7755.620,8,430.867778
6,AB-10105,14473.571,10,723.678550
7,AB-10150,966.710,5,80.559167
8,AB-10165,1113.838,8,79.559857
9,AB-10255,914.532,9,65.323714


In [83]:
# Save the cleaned DataFrame:
df.to_csv('superstore_cleaned.csv', index=False)
customer_summary.to_csv('customer_summary.csv', index=False)